pkg/agent/puller.go

```diff
@@ -1,73 +1,130 @@
 package agent
 
 import (
-	"fmt"
 	"github.com/kubeflow/kfserving/pkg/agent/storage"
+	v1 "github.com/kubeflow/kfserving/pkg/apis/serving/v1beta1"
 	"log"
 	"path/filepath"
 )
 
+type OpType string
+
+const (
+	Add    OpType = "Add"
```

Comment:
Does `Add` include `Update`? That was the initial design.
I remember you mentioned assuming that `TrainedModel` objects should be treated as immutable. 
But i disagreed. Example: 
```
TrainedModel 
name: foo
resource: 1g
storage: s3://here/1
```
if i modify resource I want to send an update command while not re-downloading. If i want to modify storage to:
`s3://there/1` and then change it back to `s3://here/1` I don't want to re-download `s3://here/1`. Example, what if I mis-type something, don't want to delete, would rather only delete up on `TrainedModel` deletion

Reply:
> `TrainedModel` objects should be treated as immutable.

I thought there was some agreement on this on the last call we had. I feel fairly strongly that it's needed, it is analogous to why pods are immutable. If you change a pod spec the pod does not reconfigure itself, a new pod with a new unique name replaces it, orchestrated via replica sets. This goes hand-in-hand with the proposal for InferenceRouters to be the vehicle for versioning and stable inferencing endpoints.

So it's important to close on this one way or the other I think since it has significant design implications.

cc @yuzisun @ellistarn @yuzliu

Reply:
I might have missed that, I am willing to be convinced otherwise :) I just think that the creation of a new model, of the same name would be common if users update their `TrainedModel`. Example: users update InferenceService when trying to patch their version, they would probably want assume the same behavior for the `TrainedModel`s.

Reply:
Right, I agree that appears kind of inconsistent with being able to directly change the `storageUri` of a single-model InferenceService. But it ultimately comes down to the fact that there isn't any indirection of the model names so there's no way that this could be implemented non-disruptively in the general case. Changing the `storageUri` like that would entail a small outage, possibly indefinite if there's some issue with the new URI/model data, which is a non-starter i.m.o.